libvast/src/partition_synopsis.cpp

```diff
@@ -34,16 +36,54 @@ void partition_synopsis::shrink() {
   }
 }
 
+// TODO: Use a more efficient data structure for rule lookup.
+std::optional<double> get_field_fprate(const index_config& config,
+                                       const qualified_record_field& field) {
+  for (const auto& [targets, fprate] : config.rules)
+    for (const auto& name : targets)
+      if (name.size()
+            == field.field_name().size() + field.layout_name().size() + 1
+          && name.starts_with(field.layout_name())
+          && name.ends_with(field.field_name()))
+        return fprate;
+  return std::nullopt;
+}
+
+double get_type_fprate(const index_config& config, const type& type) {
+  for (const auto& [targets, fprate] : config.rules) {
+    for (const auto& name : targets) {
+      if (name == ":string" && type == string_type{})
+        return fprate;
+      else if (name == ":address" && type == address_type{})
+        return fprate;
```

Comment:
Do we already have a follow-up story for generalizing this with the upcoming extractor change? Right now the logic is written per leaf in the schema, but it'd probably make more sense to have a single function that takes just an extractor and the schema in the medium term, and maps the offset of the fields to the fp-rate.

Reply:
What exactly are the relevant upcoming extractor changes you're thinking of?

There's currently no follow-up story for this, in general we've tried to keep this minimal since there was the prospect of a complete rewrite based on sketches in the near future anyways.

Reply:
Internal ID is `sc-31993`. Rough summary of the relevant bits: We unify type and field extractors and allow the resolve function on types to resolve them both into a list of offsets. That allows us to use type extractors wherever we currently use field extractors.

Reply:
I've created `sc-32362` as a follow-up.